### ISSUE TITLE
The property is not readable because it is typed "string"

### ISSUE THREAD

**Issue by guillaumesmo** [api-platform/core#3974]
**API Platform version(s) affected**: 2.6.0

**Description**  
When doing a PATCH operation with a nested collection which has typed properties, it throws an exception

**How to reproduce**  
```php
/**
 * @ApiResource(denormalizationContext={"groups"={"group:write"}})
 */
class Group {
    /**
     * @ORM\OneToMany(targetEntity=Item::class, mappedBy="group", cascade={"persist", "remove"}, orphanRemoval=true)
     * @Groups({"group:write"})
     */
    private Collection $items;
...
}

/**
 * @ApiResource
 */
class Item {
    /**
     * @ORM\Column(type="string", length=255)
     * @Groups({"group:write"})
     */
    private ?string $name;
...
}
```

```
PATCH /api/groups/1
Content-type: application/merge-patch+json
{
"items": [{"id": "/api/items/1", "name": "test"}]
}
```

```
The property "App\Entity\Item::$name" is not readable because it is typed "string". You should initialize it or declare a default value instead.

Symfony\Component\PropertyAccess\Exception\UninitializedPropertyException
in vendor/symfony/property-access/PropertyAccessor.php (line 448)
in vendor/symfony/property-access/PropertyAccessor.php -> readProperty (line 115)
in vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php -> getValue (line 563)
in vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php -> getAttributeValue (line 329)
in vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php -> denormalize (line 250)
in vendor/api-platform/core/src/Serializer/ItemNormalizer.php -> denormalize (line 70)
in vendor/symfony/serializer/Serializer.php -> denormalize (line 208)
in vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php -> denormalize (line 486)
in vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php -> denormalizeRelation (line 445)
in vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php -> denormalizeCollection (line 737)
in vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php -> createAttributeValue (line 403)
in vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php -> setAttributeValue (line 336)
in vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php -> denormalize (line 250)
in vendor/api-platform/core/src/Serializer/ItemNormalizer.php -> denormalize (line 70)
in vendor/symfony/serializer/Serializer.php -> denormalize (line 208)
in vendor/symfony/serializer/Serializer.php -> denormalize (line 144)
in vendor/api-platform/core/src/EventListener/DeserializeListener.php -> deserialize (line 106)
in vendor/symfony/event-dispatcher/Debug/WrappedListener.php -> onKernelRequest (line 117)
in vendor/symfony/event-dispatcher/EventDispatcher.php -> __invoke (line 230)
in vendor/symfony/event-dispatcher/EventDispatcher.php -> callListeners (line 59)
in vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php -> dispatch (line 151)
in vendor/symfony/http-kernel/HttpKernel.php -> dispatch (line 133)
in vendor/symfony/http-kernel/HttpKernel.php -> handleRaw (line 79)
in vendor/symfony/http-kernel/Kernel.php -> handle (line 195)
```

**Possible Solution**  
?

**Additional Context**  
this works in 2.5.9 and 2.5.10

**Comment by WhiteRabbitDE** [comment]
I think this is more of a PHP error and not APIP related. Like said in the error message, you should initalize the property.

`private ?string $name; = null`

**Comment by guillaumesmo** [comment]
Hello, I understand very well what this error stands for, but I have to disagree with the fact that this is not an api-platform issue:

- It works on versions 2.5.x, so when upgrading to a more recent minor version, I do not expect such breaking changes. Why would that not be a PHP issue in 2.5 while it is in 2.6?
- The fact that API platform is trying to read data from an uninitialized object when PATCHing an existing id smells like there is another underlying issue. Unfortunately I haven't been able to dig deeper into that.

**Comment by dunglas** [comment]
I agree that API Platform shouldn't try to access an uninitialized property. It's probably similar to https://github.com/symfony/symfony/pull/38900.

A failing test would help a lot to fix the issue.

I've the feeling that it should be fixed directly in Symfony PropertyAccess.

**Comment by guillaumesmo** [comment]
Serializer throws the exception indeed, but I think that is not the cause of the issue here. 
I wonder why API platform is trying to read a property in a non-initialized entity. Since this is a PATCH request, I expect it to load the entity, after which the properties are set (by Doctrine), then only do some Serializer read/write operations. I'm afraid this "unexpected behaviour" might also hide other issues which will later become features if left unfixed.
Also, it did work in 2.5.10, which was using Serializer as well.

I was able to reproduce the issue in the tests:
https://github.com/api-platform/core/commit/2ed2566eb3fc147d5afebc170ffcc535b0066709

Behat (PHP 7.4) (MySQL) output:
```
--- Failed steps:

001 Scenario: Update a resource with typed properties # features/main/crud.feature:93
      Then the response status code should be 200     # features/main/crud.feature:106
        Current response status code is 500, but 200 expected. (Behat\Mink\Exception\ExpectationException)

638 scenarios (634 passed, 1 failed, 3 undefined)
4239 steps (4228 passed, 1 failed, 4 undefined, 6 skipped)
6m51.50s (177.12Mb)
```

same changes in 2.5.10:
https://github.com/api-platform/core/commit/5e87c3a4280ad6bcb98e5ee8a0d401123f35cda7

Behat (PHP 7.4) (MySQL) output:
```
570 scenarios (568 passed, 2 undefined)
3788 steps (3786 passed, 2 undefined)
4m26.61s (128.76Mb)
```

**Comment by soyuka** [comment]
I think that it wasn't using the same serializer version we didn't change anything in the serialization if my memory serves well.

**Comment by guillaumesmo** [comment]
Tests were both run on symfony/serializer v5.2.1

I've reran tests on the two commits, same issue with 2.6.1 on Symfony 5.2.3, while tests pass on 2.5.10 with Symfony 5.2.3

**Comment by Tjeerd** [comment]
I can confirm the issue, after upgrading to v2.6 we are experiencing the same error on a nested entity. In our case we are using DTO's without Doctrine.

As already mentioned in https://github.com/api-platform/core/issues/3974#issuecomment-766662377, this seems to be an actual regression instead of a PHP error. In our case the workaround to default the value to null is not acceptable either:

`
final class AddressInput
{
    public DateTimeInterface $startDate;
...
}
`
Making the property `null` by default would also require it to be nullable or optional (`?DateTimeInterface`), that's not wanted. It should be a required property.

Without doing extensive digging it might be worth noting that it seems to occur when `DEEP_OBJECT_TO_POPULATE` is relevant/true.

**Comment by Tjeerd** [comment]
BTW, our case is 'just' a nested DTO, no collection:

```
final class RelationInput
{
    public ?AddressInput $address = null;
}

final class AddressInput
{
    public DateTimeInterface $startDate;
}
```

**Comment by Tjeerd** [comment]
I think the fix could/should indeed be in the Serializer:

In file `symfony/serializer/Normalizer/AbstractObjectNormalizer.php`, line 331, also catch the `UninitializedPropertyException`:

```
if ($context[self::DEEP_OBJECT_TO_POPULATE] ?? $this->defaultContext[self::DEEP_OBJECT_TO_POPULATE] ?? false) {
    try {
        $context[self::OBJECT_TO_POPULATE] = $this->getAttributeValue($object, $attribute, $format, $context);
    } catch (NoSuchPropertyException|UninitializedPropertyException $e) {
    }
}
```

But I have absolutely no idea if this 'just' fixes our usecases & does not break any other functionality.

**Comment by vuryss** [comment]
Any progress on that? I would love to remove the default null from each property and be able to use object with uninitialized properties instead.

**Comment by dshiryaev-plesk** [comment]
Workaround: create constructor for your entity class.

**Comment by vuryss** [comment]
> Workaround: create constructor for your entity class.

That's not a solution, cause I want them to be without a value until there is actually something to store. I want to be able to avoid the NULLs all over the place. Or the empty initial value too.

If the field is not set - it should simply be ignored and not throw an error, at least this sound logical - cause this is how it works without the mappings (see the symfony issue).

**Comment by guillaumesmo** [comment]
> Any progress on that?

There's not going to be any progress unless someone/you dedicates some time to work on it
I haven't had any spare time / motivation to do so recently, so I would dare to blame anyone else

In case you don't use 2.6 features (which you probably don't since you are upgrading from an older version) a good workaround is also just `composer require api-platform/core:2.5.10`

**Comment by hightem** [comment]
it is a bit a shame to have that bahaviour on core >= 2.6.0.
I am new to api platform and I have used 2.6 features that I lose if I downgrade to 2.5.10 as a workaround to the described error.
I made the choice to use core:2.6.5 and not support the patch operations on nested collections until a fix comes along, hope it will be soon

**Comment by rubencherbit** [comment]
We have the same problem...

**Comment by hightem** [comment]
@zexa it's more a workaround than a solution ... having to define a default value for all your variables is a bit heavy and more importantly not very clean imo.

**Comment by dshiryaev-plesk** [comment]
> Raised an issue in symfony for possible serializer fix: [symfony/symfony#40578](https://github.com/symfony/symfony/issues/40578)

@vuryss , thank you for your fix in Symfony. It is in Symfony 5.4 milestone now.

**Comment by Tjeerd** [comment]
Yes, thanks @vuryss 🥇 💯, with your contribution, we're able to finally switch to api-platform 2.6 and enjoy the other features!

**Comment by dant89** [comment]
@Tjeerd how did you manage to switch? We're still getting this issue and I'm not sure the fix is live in Symfony?

**Comment by Tjeerd** [comment]
> @Tjeerd how did you manage to switch? We're still getting this issue and I'm not sure the fix is live in Symfony?

I didn't switch... yet 😅 I am just happy the possibility exists as soon as Symfony 5.4 is out...
